packages/contractkit/src/wrappers/Attestations.ts

```diff
@@ -1,18 +1,201 @@
+    account: Address
+  ): Promise<ActionableAttestation[]> {
+    const phoneHash = PhoneNumberUtils.getPhoneHash(phoneNumber)
+    const expirySeconds = await this.attestationExpirySeconds()
+    const nowInUnixSeconds = Math.floor(new Date().getTime() / 1000)
+
+    const issuers = await this.contract.methods.getAttestationIssuers(phoneHash, account).call()
+    const issuerState = Promise.all(
+      issuers.map((issuer) =>
+        this.contract.methods
+          .getAttestationState(phoneHash, account, issuer)
+          .call()
+          .then(parseAttestationInfo)
+      )
+    )
+
+    // Typechain is not properly typing getDataEncryptionKey
+    const publicKeys: Promise<string[]> = Promise.all(
+      issuers.map((issuer) => this.contract.methods.getDataEncryptionKey(issuer).call() as any)
+    )
+
+    const isIncomplete = (status: AttestationState) => status === AttestationState.Incomplete
+    const hasNotExpired = (time: number) => nowInUnixSeconds < time + expirySeconds
+    const isValidKey = (key: string) => key !== null && key !== '0x0'
+
+    return zip3(issuers, await issuerState, await publicKeys)
+      .filter(
+        ([_issuer, attestation, publicKey]) =>
+          isIncomplete(attestation.attestationState) &&
+          hasNotExpired(attestation.time) &&
+          isValidKey(publicKey)
+      )
+      .map(([issuer, attestation, publicKey]) => ({
+        ...attestation,
+        issuer,
+        publicKey: publicKey.toString(),
+      }))
+  }
+
+  /**
+   * Completes an attestation with the corresponding code
+   * @param phoneNumber The phone number of the attestation
+   * @param account The account of the attestation
+   * @param issuer The issuer of the attestation
+   * @param code The code received by the validator
+   */
+  complete(phoneNumber: string, account: Address, issuer: Address, code: string) {
+    const phoneHash = PhoneNumberUtils.getPhoneHash(phoneNumber)
+    const expectedSourceMessage = attestationMessageToSign(phoneHash, account)
+    const { r, s, v } = parseSignature(expectedSourceMessage, code, issuer.toLowerCase())
+    return wrapSend(this.kit, this.contract.methods.complete(phoneHash, v, r, s))
+  }
+
+  /**
+   * Given a list of issuers, finds the matching issuer for a given code
+   * @param phoneNumber The phone number of the attestation
+   * @param account The account of the attestation
+   * @param code The code received by the validator
+   * @param issuers The list of potential issuers
+   */
+  findMatchingIssuer(
+    phoneNumber: string,
+    account: Address,
+    code: string,
+    issuers: string[]
+  ): string | null {
+    const phoneHash = PhoneNumberUtils.getPhoneHash(phoneNumber)
+    for (const issuer of issuers) {
+      const expectedSourceMessage = attestationMessageToSign(phoneHash, account)
+      try {
+        parseSignature(expectedSourceMessage, code, issuer.toLowerCase())
+        return issuer
+      } catch (error) {
+        console.log(error)
```

Comment:
Not sure about this...
do we want to hide errors? shouldn't we fail on first error?

Reply:
This is basically to find the right match in the issuer set, so the only way to figure that out is to try it which induces expected failures for issuers that are not matching